### ISSUE TITLE
[Question] Configuring VM Networking to expose IP address

### ISSUE THREAD

**Issue by BringerOD** [harvester/harvester#2291]
I have installed harvester 1.02 on 3 servers OVHCloud. 

Hardware
3 x AMD Epyc 7413 256GB Mem 2xNic

I set the VIP to the public nic IP address

Each server has a public nic and has a public IP assigned and is exposed.
The second nic is private connecting all 3 servers via a vlan.

On a vm I create, how do I assign a public IP,  or expose ports publicly?

**Comment by rodrigo14miguel** [comment]
Hi @BringerOD 
You can assign a network interface to your VM that bridges to the host management NIC (your public NIC). This VM's traffic will go out though your physical host management interface (public NIC).

Hope to have helped!

**Comment by BringerOD** [comment]
@rodrigo14miguel,  thanks.  I should have mentioned that I can create VM's that do get out to the internet,  but what I cannot do is forward a port to that vm.

Say I want to host a website on the VM.

**Comment by rodrigo14miguel** [comment]
When you say public network and private network (and regarding cloud providers), I assume you are talking about network egress accessibility to the internet, as private network cannot access the internet. Please correct me if this is not your case.
Note: I use egress as outgoing connections and ingress as incoming connections. If I make a request to the VM, to the VM this is ingress. If the VM makes a request to me, to the VM this is egress.

So you have a VM that has a network interface connected to your management network in Harvester (your public NIC on the hosts) in bridge mode. With this, you can assign your VM an IP from the public IP range (either manually or through DHCP).
I think beyond that, you would need some kind of feature from your cloud provider that redirects internet ingress traffic to your VMs public IP (and desired ports).

**Comment by BringerOD** [comment]
@rodrigo14miguel thanks again for the help.

3x Host with 2x network cards.  Each host has a nic with a public ip address assigned and 1 nic that is in a VLAN.

When I configured harvester, I chose the public network card and the public IP address that network was assigned by the cloud provider (OVH) as the VIP for each host, and then joined the other two host.  I think after further thought, I should have used the private ip address, meaning the VIP address should have been the vlan private nic, not sure.  Either way that is how its currently setup.

In harvester, I went to the Advanced -> Networks -> ClusterNetwrork:Vlan settings and enabled it.  With the default network interface as harvester-mgmt.

Then when I create the VM I used the management network in bridge mode.  The VM starts and is assigned an IP address 10.52.2.125, not sure from were though.  That VM, can get to the internet, or egress to the internet.  

Here is what the network card was assigned. 
![image](https://user-images.githubusercontent.com/7971305/170524997-c425085a-a2f3-43e3-ad68-c7b4efdaebbe.png)

So are you saying,  configure the VM,  with a public IP address?  Attempting that now.

**Comment by rodrigo14miguel** [comment]
@BringerOD thats odd. That IP is strange. I was expecting bridge to work with the management interface IP range... at least thats what makes sense from bridge concept. Yet it is a kubernetes cluster after all.

I have seen somewhere that management network is normally used to have management network access but the bridge concept should have worked anyway, spatially when single NIC deployment is a thing. I will try to make some more test when I can to see how to make it work if possible.

**Comment by BringerOD** [comment]
@rodrigo14miguel again, thanks for your attention on this.  I am just really stuck on how to proceed.  I am ok with networking, not an expert, and trying to get this up and running has been challenging.  Here is a support bundle

https://www.dropbox.com/s/fv5ddyfs9koym6q/supportbundle_bba4fe6c-1669-4256-a730-b737e3ecd538_2022-05-26T19-22-08Z.zip?dl=0

If that helps.

**Comment by rodrigo14miguel** [comment]
Hi @BringerOD 
I have deployed a single node Harvester 1.0.2 with 2 NICs to test this out:
- NIC 1 for MGMT network
- NIC 2 for VLAN trunk

So here follows my findings:
- assigning MGMT network to VM interface as bridge -> since the MGMT interface is bonded, Harvester creates an internal bridge network with an internal IP. This your VM getting an IP of 10.52.2.x. This option wont solve your issue since your VM interface is in a private network. Even if you assign a static public IP with a public gateway to your VM, it will just be an IP sitting in an unknown network for him (no routing or switching).
- Setting NIC 2 as default VLAN network and creating a VLAN tagged network -> Harvester creates a bridge network with the physical interface of your host. In other terms, it is bridging with the outside network. With this, I'm able to assign an outside IP to my VM, pinging the outside world from the VM and pinging the VM from an outside machine.

I haven't tried this 3rd method as I think it would fail:
1.  setting the MGMT network as trunk with my harvester management VLAN as native
2. assigning the MGMT network as default VLAN network in harvester
3. creating a VLAN tagged network (with a tag other than the MGMT one)

The reason I haven't tried this is bc from what I see after assigning the MGMT network as default VLAN network is that Harvester creates a bridge network assigned to the MGMT bond. I assume this will produce similar results as assigning the MGMT network directly to the VM in bridge mode, as it will probably create a private network for that (10.x.x.x I think). You would still be able to reach the internet from your VM but wont be able to assign it a public IP (or at least make it reachable from outside of harvester). Just assumptions.

In other words, I cannot help you put your setup to work as is...
Alternatively, you could have your public VLAN network configured in your second NIC's trunk and use the second NIC as your default VLAN network. After creating a VLAN network in harvester tagged that public VLAN, you can use it in your VM and assign it a public IP, reachable from the outside.

Hope this helps!

**Comment by spatialy** [comment]
Hi @rodrigo14miguel 

We are stuck trying to get the second option to work, we have a deployment where each node has 2 NICs, the second NIC bond we create manually as per the documentation but each time we create a network in harvester with the VLAN dedicated NIC the DHCP validation fails and the VM doesn't get any external assigned IP from the DHCP.

Can you help us with more details on how you deploy the second scenario?

Best

**Comment by rodrigo14miguel** [comment]
Hi @spatialy 

In your Networks tab in Harvester's UI, does this network that you speak of show any yellow balloon saying "DHCP failed" (or something like that) to the right?

**Comment by spatialy** [comment]
Hi @rodrigo14miguel, yes we see the ballon with the message … we don’t know where else to check … if you think a support bundle can help we can upload one

**Comment by rodrigo14miguel** [comment]
Hi @spatialy,

I remember encounter a similar problem. I had a DHCP server external to Harvester.
When creating a VLAN network in Harvester, for the "route" options of the VLAN network, when I specified my DHCP server IP (that was on the same VLAN), the VLAN network in Harvester would display that yellow balloon and it would not work.
 
To make it work, I left the "route"'s DHCP IP section empty for the VLAN network configuration. The VLAN network would then show as ready (maybe its a bug or something, don't really know what that field is for after seeing this behavior). When assigning a VM with this VLAN network, it would then grab an IP from my external DHCP IP pool.

Here is an image of the docs that is verry similar to what I had more or less (external DHCP server on the same VLAN network as the one created in Harvester)(https://docs.harvesterhci.io/v1.0/networking/harvester-network/):
![alt text](https://docs.harvesterhci.io/v1.0/networking/assets/vlan-case.png "Title")

Hope this helps!
Let me know if it worked for you

**Comment by spatialy** [comment]
Yes same scenario here, let try to get the field blank and see.

Last question, you use 1 or 2 NIC? We use 1 dedicated NIC for each network … docs and schematics don’t get clear on this beside you need to bond the second NIC.

Let you know how go with your same steps.

Best

**Comment by rodrigo14miguel** [comment]
I used 2 NICs:

- NIC 1 for Harvester management traffic and VIP (access UI also). The wizard creates a bond with a single network interface.
- NIC 2 for VM traffic, multiple VLANs. No bonding configuration needed, just chose it as default VLAN network interface.

If you are doing a baremetal installation with the ISO, you can do interface bonding for redundancy. In the installation process you can bond several interfaces for the management network. If your upstream switch supports it, I would recommend use LACP (option 802 something something in the installation's load balancing option, don't remember exactly).
For the second bond (the one for VM VLAN traffic), you would need to ssh into the Harvester hosts and do the bonding by hand as there is no way to do it though the UI or installation wizard (Harvester 1.0.2). I haven't done this second part myself but I have seen it somewhere in the documentation.

For example, if you have 4 network interfaces, you could do:
 - bond 1 (interface 1 and 2) - harvester management traffic, configured though installation wizard
 - bond 2 (interface 3 and 4) - VM VLAN traffic, configured though linux operating system configuration

As I said, I strongly recommend LACP for the bonds as it provides active-active load balancing, if your upstream switch supports it.

**Comment by w13915984028** [comment]
@BringerOD 

Please check if following link helps, thanks.

Re-use management network to expose VM to outside world.

https://github.com/w13915984028/harvester-develop-summary/blob/main/connect-vm-from-mgmt-network.md

**Comment by spatialy** [comment]
Hi @w13915984028, glad this catch your attention.

We don't need to access from outside the VMs, as we plan to expose services using an IPPool as per docs.

What we are troubling is on getting the second dedicated NIC to work for the VLAN ... we try the suggestions of the documentation and really do a test with bonding, no bonding, DHCP IP defined and not, as well a direct route on Harester network configs.

The only way we get the VMs an IP is when get them using the MGMT-NET NIC.

I can confirm the base nodes are getting IPs from the DHCP over a tagged VLAN for each physical interface.

If help here is a support bundle.
[supportbundle_ff02cf6b-d763-4b2f-b6ff-ff777adbdd42_2022-06-21T16-04-23Z.zip]
(https://github.com/harvester/harvester/files/8957445/supportbundle_ff02cf6b-d763-4b2f-b6ff-ff777adbdd42_2022-06-21T16-04-23Z.zip)

Our goal is to make Harvester use the dedicated NIC for VLAN.

Best

**Comment by w13915984028** [comment]
@spatialy  Please check following doc, I tried with dedicated VLAN network in nested virtualization. You can map it to your network setting, and even trouble shooting via the related path. When it is sitll not working, please depict your network topology, and list the related setting in your SWITCH and bare metal machines (are you using bare metal machine ?).

https://github.com/w13915984028/harvester-develop-summary/blob/main/dive-into-harvester-vlan-network-in-nested-virtualization.md

**Comment by BringerOD** [comment]
So we ended up getting this all to work using only one nic.  Mainly because the public nic was really only designed, by the provider OVH, for public access.  Since the harvester cluster is private,  we ended up only using the private nic.  

We finally did get it all working.  We are going to post a case study here pretty soon on how we did it.

- 3 Node Harvester Cluster
- pFSense server for firewall, DHCP. HAPROXY
- Rancher Cluster

We got MetalLB and Traefik working on our Kubes as well,  exposed to the internet, fault tolerant, and load balanced.

**Comment by tjjh89017** [comment]
I will close this ticket.
feel free to reopen it.

**Comment by spatialy** [comment]
yes, it showed the ballon, i can upload a support bundle if needed

**Comment by fachex** [comment]
> So we ended up getting this all to work using only one nic. Mainly because the public nic was really only designed, by the provider OVH, for public access. Since the harvester cluster is private, we ended up only using the private nic.
> 
> We finally did get it all working. We are going to post a case study here pretty soon on how we did it.
> 
> * 3 Node Harvester Cluster
> * pFSense server for firewall, DHCP. HAPROXY
> * Rancher Cluster
> 
> We got MetalLB and Traefik working on our Kubes as well, exposed to the internet, fault tolerant, and load balanced.

Hello, I am very interested to know how you made it to work on OVH. I will appreciate any documentation you might have. Thank you!

**Comment by xeruf** [comment]
Similar issue I think, explanation of the solution would really help :) https://github.com/harvester/harvester/issues/4129
